### PR TITLE
Add some basic QUnit tests for Morebits functions

### PR DESCRIPTION
**Pull request by Amorymeltzer**
tl;dr: QUnit tests (#309) for some Morebits functions `test/tests/`, run via `npm run test`, part of CI.  REPL to play around with.  cc @MusikAnimal and @siddharthvp 

Makes use of jsdom (and jquery), as well as

Makes use of jsdom, as well as jQuery (which is added @3.5.1 rather than MediaWiki's @3.4.1 since the differences don't matter here (https://phabricator.wikimedia.org/T250068)).  In the future jquery-ui could be added (the MediaWiki version (1.9.2) is no longer available (https://www.npmjs.com/package/jquery-ui)).

This a little shim for mediawiki functions and some false data for quick testing essentially creates a REPL, which we load before the QUnit tests.  Each class has its own file, and are all fairly basic right now but should hopefully catch some things down the line.  Nearly all the issues with `Morebits.date` have been browser-specific (implementations of `Date.parse`, polluted prototypes, etc.), which this won't help with.  But a good start!  The tests have been added to GitHub actions.

The mediawiki functions and data are a limited set (see files for list) as needed for the unit tests, although there are extra `mw.util` functions we don't need at the moment.  In the future, it would be nice to set this up to have multiple page/action/user examples.

Per #1142, done away with the `dev/` directory; moved test files into test folder, moved `sync.pl` into `scripts/` folder.

If you want to use the testable REPL environment for morebits, just run `require('path/to/mb_repl')`.  This essentially does:

```javascript
const jsdom = require('jsdom');
const { JSDOM } = jsdom;
global.window = (new JSDOM()).window;
global.document = window.document;

global.HTMLFormElement = window.HTMLFormElement;

global.jQuery = require('jquery');
```

And then loads in `mw_shim`, `mw_config_data`, and finally `morebits.js`, e.g. (from the top of the directory):

```javascript
global.mw = require('./test/mw_shim').mw;
require('./test/mw_config_data');
require('./morebits.js');
global.Morebits = window.Morebits;
```

----

Currently has basic tests for:
- toplevel properties and functions
- `Morebits.date`
- `Morebits.string`
- `Morebits.array`
- `Morebits.wikitext` and `Morebits.wikitext.page`
- `Morebits.batchOperation`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,12 +6,16 @@
 	"scripts": {
 		"lint": "eslint .",
 		"lint:fix": "eslint . --fix",
-		"patchtest": "node dev/patch-test.js",
-		"sync": "perl dev/sync.pl",
+		"patchtest": "node test/patch-test.js",
+		"sync": "perl scripts/sync.pl",
+		"test": "qunit --require ./test/mb_repl.js ./test/tests/*",
```
Comment (Amorymeltzer) [inline package.json:11]: This could arguably be `lint && test`

File: test/tests/morebits.wikitext.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+QUnit.module('Morebits.wikitext');
+QUnit.test('parseTemplate', assert => {
+	var makeTemplate = function(data) {
+		var template = '{{' + data.name;
+		Object.keys(data.parameters).forEach(function(key) {
+			// Necessary until fixing a bug in parseTemplate
```
Comment (Amorymeltzer) [inline test/tests/morebits.wikitext.js:6]: #1214

**Comment by siddharthvp** [comment]
Aren't there licensing issues with copying large chunks of MW source code into the repo? Neither CC-BY-SA nor GFDL are GPL-compatible.

**Comment by Amorymeltzer** [comment]
> Aren't there licensing issues with copying large chunks of MW source code into the repo? Neither CC-BY-SA nor GFDL are GPL-compatible.

Well, none of it is going on the wiki, so that's not a concern.  We're not explicitly licensed here (#252, #253) (although obviously all on-wiki code is implicitly cc-by-sa/gfdl), but it's a good point, I'll explicitly include the header on that one file.

File: test/tests/morebits.array.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+QUnit.module('Morebits.array');
+QUnit.test('chunk', assert => {
+	assert.deepEqual(Morebits.array.chunk([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 3), [[1, 2, 3], [4, 5, 6], [7, 8, 9], [10]], '10 over 3');
+	assert.deepEqual(Morebits.array.chunk([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 42), [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]], 'Size > length');
+	assert.deepEqual(Morebits.array.chunk([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]), [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]], 'No size');
+	// assert.throws(Morebits.array.chunk(42), 'throws');
```
Comment (Krinkle) [inline test/tests/morebits.array.js:6]: In case you were wondering, using `throws()` takes the statement as a code block. Right now it would execute outside the test and not possible for QUnit to check it. (See <https://api.qunitjs.com/assert/throws/>.)

Like so:

```js
assert.throws(function() {
  Morebits.array.chunk(42);
}, 'No array');
```

Or pick one of the shorter styles that JavaScript allows:

```js
assert.throws(
  () => { Morebits.array.chunk(42); },
  'No array'
);

assert.throws(() => Morebits.array.chunk(42), 'No array');
```

File: test/mw_shim.js [hunk 4]
```diff
@@ -0,0 +1,849 @@
...
+ */
+mw.log = console && Function.prototype.bind.call(console.log, console);
+mw.log.warn = console && console.warn ?
+	Function.prototype.bind.call(console.warn, console) :
+	function(){};
+
+/**
+ * mw.Map
+ * Needed for various things, in particular mw.config
+ * https://doc.wikimedia.org/mediawiki-core/master/js/source/mediawiki.html#mw-Map-method-constructor
```
Comment (Krinkle) [inline test/mw_shim.js:36]: These are not stable links for source code.

Use something like <https://gerrit.wikimedia.org/g/mediawiki/core/+/1.35.0/resources/src/startup/mediawiki.js> or <https://github.com/wikimedia/mediawiki/blob/ea586cfa2e/resources/src/startup/mediawiki.js> instead.

When browsing GitHub, you can press `y` to get a stable permalink.

File: .github/workflows/mb_funcs.yml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+name: Morebits function tests
+
+on: [push, pull_request]
```
Comment (siddharthvp) [inline .github/workflows/mb_funcs.yml:3]:
```suggestion
on:
  push:
    paths:
    - 'morebits.js'
  pull_request:
    paths:
    - 'morebits.js'
```

No need to trigger on every PR, just the ones that change morebits. Per https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#example-ignoring-paths

File: test/mb_repl.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+ */
+
+// Consistent environment stuff (mw, window/document via jsdom, jQuery)
+require(__dirname + '/mb_repl_head.js');
+
+// Load mw.config data
+require(__dirname + '/mw_config_data.js');
+
+require(__dirname + '/../morebits.js');
+global.Morebits = window.Morebits;
```
Comment (siddharthvp) [inline test/mb_repl.js:13]: Neat, I had heard of `globalThis` but didn't know `global` was a thing!

File: test/mw_config_data.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
+// Includes only what's needed for morebits, util, and Title
+mw.config.set({
+	// Needed by Morebits
+	pageTriageNamespaces: [0, 2, 118],
+	wgArticleId: 18866,
+	wgIsRedirect: false,
+	wgPageName: 'Macbeth',
+	wgRestrictionEdit: ['autoconfirmed'],
+	wgUserGroups: ['interface-admin', 'oversight', 'patroller', 'reviewer', 'sysop', '*', 'user', 'autoconfirmed'],
```
Comment (siddharthvp) [inline test/mw_config_data.js:9]: Ooof!

File: test/mw_shim.js [hunk 8]
```diff
@@ -0,0 +1,849 @@
...
+ * user.tokens (nonfunctional, only used in Morebits.wiki.page)
+ * Title (only used in Morebits.wiki.page)
+ */
+
+var mw = {};
+
+/**
+ * mw.log
+ * We just need the `warn` method, but might as well mw.log = console.log
+ * node repl doesn't have console.warn
```
Comment (siddharthvp) [inline test/mw_shim.js:25]: ? node repl does have it.

File: test/tests/morebits.date.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
+QUnit.module('Morebits.date');
+var now = Date.now();
+var timestamp = '09:31, 17 April 2020 (UTC)';
+QUnit.test('Construction', assert => {
+	assert.equal(new Morebits.date().getTime(), new Date().getTime(), 'Basic constructor');
```
Comment (siddharthvp) [inline test/tests/morebits.date.js:5]: This failed for me the first time I ran it (off by one 1 millisecond), but worked fine on every later run.

File: test/tests/morebits.wikitext.js (hunk 2 above)
Comment (siddharthvp) [reply test/tests/morebits.wikitext.js:6]: You can write the test for the ideal behaviour here and disable it using `QUnit.skip`, to be enabled later. Writing flawed tests just to get them to pass defeats the point of testing.

File: test/tests/morebits.batchOperation.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+var chunkLength = 2;
+batch.setOption('chunkSize', chunkLength);
+QUnit.test('Run', assert => {
+	var all = '';
+	var runFunc = function(page) {
+		all += page;
+		batch.workerSuccess();
+	};
+	batch.run(runFunc);
+	assert.equal(all, pageList.join(''), 'batch run');
```
Comment (siddharthvp) [inline test/tests/morebits.batchOperation.js:17]: `assert.equal`, here and everywhere, does comparisons with `==`. Should be avoided in favour of `assert.strictEqual`. This is why Qunit is annoying.

File: test/mw_config_data.js [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+	wgUserName: 'Amorymeltzer',
+
+	// Needed by mw.util
+	wgScript: '/w/index.php',
+	wgArticlePath: '/wiki/$1',
+	wgScriptPath: '/w',
+
+	// Needed by mw.Title
+	wgFormattedNamespaces: {
+                '-2': 'Media',
```
Comment (siddharthvp) [inline test/mw_config_data.js:19]: indentations are messed up in this file.

**Review by siddharthvp — COMMENTED** [review]
This looks pretty good!

Regarding the directory structure, don't do `/test/tests`. Put the mb_* and mw_* files in `/test/mocking`, and the actual tests in `/test`.

File: .github/workflows/mb_funcs.yml (hunk 5 above)
Comment (Amorymeltzer) [reply .github/workflows/mb_funcs.yml]: Dope!

File: test/mw_config_data.js (hunk 11 above)
Comment (Amorymeltzer) [reply test/mw_config_data.js]: s/spaces/tabs/

File: test/tests/morebits.array.js (hunk 3 above)
Comment (Amorymeltzer) [reply test/tests/morebits.array.js]: Awesome, thanks!  Done!

File: test/tests/morebits.batchOperation.js (hunk 10 above)
Comment (Amorymeltzer) [reply test/tests/morebits.batchOperation.js]: Done!

File: test/tests/morebits.date.js (hunk 9 above)
Comment (Amorymeltzer) [reply test/tests/morebits.date.js]: I initially had this with ISOString for exactly this reason (some browsers fudge milliseconds for security reasons), so maybe that would be better?

File: test/mb_repl.js (hunk 6 above)
Comment (Amorymeltzer) [reply test/mb_repl.js:13]: Like 70% of the effort here was annoying `global`/`window`/`global.window` shit.  But global is the repl's speciality since there ain't no window

File: test/mw_shim.js (hunk 8 above)
Comment (Amorymeltzer) [reply test/mw_shim.js]: Dunno WTF that was, I must've been overwriting it when first testing...

File: .github/workflows/mb_funcs.yml (hunk 5 above)
Comment (Amorymeltzer) [reply .github/workflows/mb_funcs.yml]: Mmm, what about adding `- 'test/**`?  Might as well run the tests if tests change...

**Comment by Amorymeltzer** [comment]
> Regarding the directory structure, don't do `/test/tests`. Put the mb_* and mw_* files in `/test/mocking`, and the actual tests in `/test`.

Fair, although I did this so as to fit all the other stuff in there (patch-test, #1215, etc.).  I don't *think* QUnit will touch anything by default, but this seemed cleaner.

File: .github/workflows/mb_funcs.yml (hunk 5 above)
Comment (siddharthvp) [reply .github/workflows/mb_funcs.yml]: Yep, was about to say that!

File: test/tests/morebits.date.js (hunk 9 above)
Comment (Amorymeltzer) [reply test/tests/morebits.date.js]: I suppose we could lop off the final digit, since we don't really care.  It's reliably under 0.1% of times for me, so I'm fine leaving as-is if you are.

**Comment by Amorymeltzer** [comment]
In the interest of just moving forward, done
